alf/algorithms/dqn_algorithm.py

```diff
@@ -0,0 +1,192 @@
+                itself.
+            critic_loss_ctor: a critic loss
+                constructor. If ``None``, a default ``OneStepTDLoss`` will be used.
+            debug_summaries (bool): True if debug summaries should be created.
+            name (str): The name of this algorithm.
+        """
+        self._rollout_epsilon_greedy = as_scheduler(rollout_epsilon_greedy)
+        # Disable alpha learning:
+        alpha_optimizer = AdamTF(lr=0)
+
+        super().__init__(
+            observation_spec=observation_spec,
+            action_spec=action_spec,
+            reward_spec=reward_spec,
+            actor_network_cls=None,
+            critic_network_cls=None,
+            q_network_cls=q_network_cls,
+            # Do not use entropy reward:
+            use_entropy_reward=False,
+            num_critic_replicas=num_critic_replicas,
+            env=env,
+            config=config,
+            critic_loss_ctor=critic_loss_ctor,
+            # Allow custom optimizer for q_network:
+            critic_optimizer=q_optimizer,
+            alpha_optimizer=alpha_optimizer,
+            debug_summaries=debug_summaries,
+            name=name)
+        assert self._act_type == ActionType.Discrete
+
+    # Copied and modified from sac_algorithm (discrete actions).
+    def _predict_action(self,
+                        observation,
+                        state: DqnActionState,
+                        epsilon_greedy=None,
+                        eps_greedy_sampling=False):
+        new_state = DqnActionState()
+        critic_network_inputs = (observation, None)
+        q_values, critic_state = self._compute_critics(
+            self._critic_networks, *critic_network_inputs, state.critic)
+        new_state = new_state._replace(critic=critic_state)
+
+        # NOTE: This block departs from SAC:
+        size = q_values.shape  # [B, actions]
+        if eps_greedy_sampling:
+            # Epsilon greedy for rollout or evaluation.
+            rand_act_prob = epsilon_greedy / size[-1]
+            probs = torch.ones_like(q_values) * rand_act_prob
+            if epsilon_greedy >= 1:
+                # Uniform random action distribution
+                greedy_act_prob = rand_act_prob
+            else:
+                # Epsilon greedy
+                greedy_act_prob = 1 - epsilon_greedy + rand_act_prob
+        else:
+            # Greedy for train_step to obtain target value from target network.
+            # The greedy action here is the maximizer of q values, and will be used
+            # to obtain target value using the target network.
+            probs = torch.zeros_like(q_values)
+            greedy_act_prob = 1
+        greedy_action = torch.argmax(q_values, dim=-1)
+        probs[(torch.arange(size[0]), greedy_action)] = greedy_act_prob
+        action_dist = td.Categorical(probs=probs)
+        action = dist_utils.epsilon_greedy_sample(action_dist, eps=0.)
```

Comment:
This place should be eps=1? For training, it produces the same result with eps=0 because of one-hot. For rollout and evaluation, it does sampling, requiring eps=1.

Reply:
Good point, done.